api/internal/core/server/server.go

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package server
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/apisix/manager-api/internal/conf"
+	"github.com/apisix/manager-api/internal/log"
+	"github.com/apisix/manager-api/internal/utils"
+)
+
+// server is the manager of Manager API, which is responsible for managing the life cycle of Manager API, including initialization, start, stop and so on
+type server struct {
+	server    *http.Server
+	serverSSL *http.Server
+	options   *Options
+}
+
+type Options struct {
+	ForceStart bool // force start new instance
+}
+
+// NewServer Create a server manager
+func NewServer(options *Options) (*server, error) {
+	return &server{options: options}, nil
+}
+
+func (s *server) Start() error {
+	// initialize server
+	err := s.init()
+	if err != nil {
+		return err
+	}
+
+	// write daemon pid file
+	err = s.writePID()
+	if err != nil {
+		return err
+	}
+
+	// print server info to stdout
+	s.printInfo()
+
+	// For internal error handling across multiple goroutines.
+	errSig := make(chan error, 1)
+
+	// start HTTP server
+	log.Infof("The Manager API is listening on %s", s.server.Addr)
+	go func() {
+		if err := s.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			log.Errorf("listen and serv fail: %s", err)
+			errSig <- err
+		}
+	}()
+
+	// start HTTPs server
+	if conf.SSLCert != "" && conf.SSLKey != "" {
+		go func() {
+			err := s.serverSSL.ListenAndServeTLS(conf.SSLCert, conf.SSLKey)
+			if err != nil && err != http.ErrServerClosed {
+				log.Errorf("listen and serve for HTTPS failed: %s", err)
+				errSig <- err
+			}
+		}()
+	}
+
+	// handle HTTP(s) server error
+	select {
+	case err := <-errSig:
```

Comment:
Frankly, the way to detect whether these servers were started is not reliable, as it depends on the behavior the go runtime, like the scheduler. These two goroutines might not get the change to run when the `select` statement is run.

Reply:
Please forgive me for not knowing the relevant content of go runtime. Can you talk more deeply or have related document content?
I search about some related questions, and the solution given is just like what these codes do now.

> https://stackoverflow.com/questions/26090301/run-both-http-and-https-in-same-program/29468115

Reply:
Just like I said:

> These two goroutines might not get the change to run when the select statement is run.

And since the `select` statement is non-block, the select run happens before the two servers try to send error message (if errors really occurred).

Reply:
Yes, here we can't get err absolutely even if the error occurs.
Maybe we can return a channel in this function.

Reply:
Yes, I understand the problems here.

Therefore, I try to change the `default case` to a 5 seconds timer. The listen operation of most devices should be completed within 5 seconds, so the current select will block start for 5 seconds, but it will not affect the HTTP server in goroutine. If the error of `Listen` does not return within 5 seconds, it is considered that no error has occurred and the timer returns `nil` error.

Reply:
**Update:** Yes, I have modified it in this way.